board/safety/safety_ford.h

```diff
@@ -1,112 +1,227 @@
     }
-    brake_pressed_prev = brake_pressed;
-  }
 
-  // exit controls on rising edge of gas press
-  if (addr == 0x204) {
-    gas_pressed = ((GET_BYTE(to_push, 0) & 0x03U) | GET_BYTE(to_push, 1)) != 0U;
-    if (!unsafe_allow_gas && gas_pressed && !gas_pressed_prev) {
-      controls_allowed = 0;
+    // Update brake pedal and cruise state
+    if (addr == MSG_ENG_BRAKE_DATA) {
+      // Signal: BpedDrvAppl_D_Actl
+      brake_pressed = (GET_BYTE(to_push, 0) & 0x30U);
+
+      // Signal: CcStat_D_Actl
+      unsigned int cruise_state = (GET_BYTE(to_push, 1) & 0x07U);
+      bool cruise_engaged = (cruise_state == 0U) || (cruise_state == 3U);
+
+      // Enter controls on rising edge of stock ACC, exit controls if stock ACC disengages
+      if (cruise_engaged && !cruise_engaged_prev) {
+        controls_allowed = true;
+      }
+      if (!cruise_engaged) {
+        controls_allowed = false;
+      }
+      cruise_engaged_prev = cruise_engaged;
     }
-    gas_pressed_prev = gas_pressed;
-  }
 
-  if ((safety_mode_cnt > RELAY_TRNS_TIMEOUT) && (bus == 0) && (addr == 0x3CA)) {
-    relay_malfunction_set();
+    // Check if IPMA is online
+    bool is_lkas_msg = (addr == MSG_LANE_ASSIST_DATA1)
+                    || (addr == MSG_LATERAL_MOTION_CONTROL)
+                    || (addr == MSG_IPMA_DATA);
+    generic_rx_checks(is_lkas_msg);
   }
-  return 1;
-}
 
-// all commands: just steering
-// if controls_allowed and no pedals pressed
-//     allow all commands up to limit
-// else
-//     block all commands that produce actuation
+  return valid;
+}
 
 static int ford_tx_hook(CANPacket_t *to_send) {
-
   int tx = 1;
   int addr = GET_ADDR(to_send);
+  bool violation = false;
 
-  // disallow actuator commands if gas or brake (with vehicle moving) are pressed
-  // and the the latching controls_allowed flag is True
-  int pedal_pressed = brake_pressed_prev && vehicle_moving;
-  bool unsafe_allow_gas = unsafe_mode & UNSAFE_DISABLE_DISENGAGE_ON_GAS;
-  if (!unsafe_allow_gas) {
-    pedal_pressed = pedal_pressed || gas_pressed_prev;
+  if (!msg_allowed(to_send, FORD_TX_MSGS, FORD_TX_LEN)) {
+    tx = 0;
   }
-  bool current_controls_allowed = controls_allowed && !(pedal_pressed);
-
-  // STEER: safety check
-  if (addr == 0x3CA) {
-    if (!current_controls_allowed) {
-      // bits 7-4 need to be 0xF to disallow lkas commands
-      if ((GET_BYTE(to_send, 0) & 0xF0U) != 0xF0U) {
-        tx = 0;
-      }
-    }
+
+  // Safety check for Lane_Assist_Data1 action
+  if (addr == MSG_LANE_ASSIST_DATA1) {
+    // Do not allow steering using Lane_Assist_Data1 (Lane-Departure Aid).
+    // This message must be sent for Lane Centering to work, and can include
+    // values such as the steering angle or lane curvature for debugging,
+    // but the action must be set to zero.
+    unsigned int action = (GET_BYTE(to_send, 0) & 0xE0U) >> 5;
+    violation |= (action != 0U);
```

Comment:
I've added this check to make sure the action is set to zero on one of the steering commands. It's required to send this along with the lane centering command, otherwise it doesn't work, because at the moment we block the stock message.

We would want to block it, I think, so that the car's Lane Keeping Aid does not interfere with Openpilot. Maybe we should only block it when OP is engaged and pass-thru the action otherwise? Does the fwd_hook already do that for us and the tx_hook only blocks messages OP sends?

Reply:
Short version: Panda doesn't try and shouldn't try to make factory LKAS work when openpilot is disengaged.

Long version: You would have to solve a surprising number of technical issues to make switching back and forth work reliably and safely. If you're thinking "why couldn't we just X", I assure you not only be there dragons, but dragons hiding behind the dragons. After that, there's sort of an overall end-user experience design decision to be made, the look and feel of the openpilot experience, and comma have made the decision that either openpilot is driving or the driver is driving. I happen to agree. Some disagree, quite vehemently. It's one of those things that will descend into holy flame wars exceeded only by holy flame wars about sound design. It's okay to tinker with passthrough in a fork if you really want, but the version upstreamed to comma should not attempt passthrough of factory LKAS steering commands.

Reply:
I'm happy to agree with upstream for now 😅

So, blocking all stock LKAS commands is the correct approach?